engine/runtime/src/main/java/org/enso/interpreter/runtime/data/vector/ArrowArrayBuilder.java

```
@@ -0,0 +1,239 @@
+package org.enso.interpreter.runtime.data.vector;
+
+import com.oracle.truffle.api.CompilerDirectives;
+import com.oracle.truffle.api.dsl.Cached;
+import com.oracle.truffle.api.interop.InteropLibrary;
+import com.oracle.truffle.api.interop.InvalidArrayIndexException;
+import com.oracle.truffle.api.interop.UnsupportedMessageException;
+import com.oracle.truffle.api.interop.UnsupportedTypeException;
+import com.oracle.truffle.api.library.CachedLibrary;
+import com.oracle.truffle.api.library.ExportLibrary;
+import com.oracle.truffle.api.library.ExportMessage;
+import com.oracle.truffle.api.nodes.Node;
+import org.enso.interpreter.dsl.Builtin;
+import org.enso.interpreter.node.expression.builtin.interop.syntax.HostValueToEnsoNode;
+import org.enso.interpreter.runtime.EnsoContext;
+import org.enso.interpreter.runtime.data.EnsoObject;
+import org.enso.interpreter.runtime.data.Type;
+import org.enso.interpreter.runtime.error.Warning;
+import org.enso.interpreter.runtime.error.WarningsLibrary;
+import org.enso.interpreter.runtime.error.WithWarnings;
+import org.enso.interpreter.runtime.library.dispatch.TypesLibrary;
+
+@ExportLibrary(InteropLibrary.class)
+@ExportLibrary(TypesLibrary.class)
+@Builtin(pkg = "mutable", stdlibName = "Standard.Base.Data.ArrowArrayBuilder.ArrowArrayBuilder")
+public abstract class ArrowArrayBuilder implements EnsoObject {
+
+  protected boolean sealed = false;
+
+  public void seal() {
+    sealed = true;
+  }
+
+  @ExportMessage
+  boolean hasArrayElements() {
+    return true;
+  }
+
+  @ExportMessage
+  boolean isArrayElementModifiable(long index) {
+    return true;
+  }
+
+  @ExportMessage
+  boolean isArrayElementReadable(long index) {
+    return true;
+  }
+
+  @ExportMessage
+  long getArraySize() {
+    throw CompilerDirectives.shouldNotReachHere();
+  }
+
+  @ExportMessage
+  Object readArrayElement(long index)
+      throws UnsupportedMessageException, InvalidArrayIndexException {
+    throw CompilerDirectives.shouldNotReachHere();
```

Please throw `UnsupportedMessageException` which is an expected interop exception instead of _panic exception_ like `shouldNotReachHere` which aborts execution altogether.

@radeusgd wrote:
> I think readArrayElement should be consistent with isArrayElementReadable

consistency is preferred, but there is no guarantee the callers of the `InteropLibrary` will honor the consistency! There is code in Truffle which doesn't bother checking `isXyz` and directly calls `xyz` catching the `UnsupportedException`. Such a use-case is considered valid.

---

In that case should [Vector](https://github.com/enso-org/enso/blob/develop/engine/runtime/src/main/java/org/enso/interpreter/runtime/data/vector/Vector.java#L49) also be changed? I tried to stick close to similar implementations.

---

Yes, it can be changed to not be a _misleading example_ to follow. However [Vector](https://github.com/enso-org/enso/blob/develop/engine/runtime/src/main/java/org/enso/interpreter/runtime/data/vector/Vector.java#L49) isn't the same case as you have here!

Please note that `Vector` is abstract shared class and all the real subclasses _override this message_ (but for example `Generic` does it with [completely different signature](https://github.com/enso-org/enso/blame/develop/engine/runtime/src/main/java/org/enso/interpreter/runtime/data/vector/Vector.java#L222) from Java perspective). The method in question should be `abstract` ideally, but then either `javac` or Truffle DSL complains when the method is made so.

Created https://github.com/enso-org/enso/pull/9312/commits/3a431a362be772703c542e704c5661e1cb08655e to consider some alternatives - #9312